src/js/stores/MesosStateStore.js

```diff
@@ -90,79 +49,51 @@ class MesosStateStore extends GetSetBaseStore {
       listenAlways: true
     });
 
-    this.dispatcherIndex = AppDispatcher.register(payload => {
-      if (payload.source !== ActionTypes.SERVER_ACTION) {
-        return false;
-      }
-
-      var action = payload.action;
-      switch (action.type) {
-        case ActionTypes.REQUEST_MESOS_STATE_SUCCESS:
-          this.processStateSuccess(action.data);
-          break;
-        case ActionTypes.REQUEST_MESOS_STATE_ERROR:
-          this.processStateError(action.xhr);
-          break;
-        case ActionTypes.REQUEST_MESOS_STATE_ONGOING:
-          this.processOngoingRequest();
-          break;
-      }
-
-      return true;
+    METHODS_TO_BIND.forEach(method => {
+      this[method] = this[method].bind(this);
     });
-
-    VisibilityStore.addChangeListener(
-      VISIBILITY_CHANGE,
-      this.onVisibilityStoreChange.bind(this)
-    );
   }
 
   addChangeListener(eventName, callback) {
     this.on(eventName, callback);
 
-    if (this.shouldPoll()) {
-      startPolling();
+    if (this.shouldSubscribe()) {
+      this.subscribe();
     }
   }
 
   removeChangeListener(eventName, callback) {
     this.removeListener(eventName, callback);
 
-    if (!this.shouldPoll()) {
-      stopPolling();
+    if (!this.shouldSubscribe()) {
+      this.unsubscribe();
     }
   }
 
-  onVisibilityStoreChange() {
-    if (!VisibilityStore.isInactive() && this.shouldPoll()) {
-      startPolling();
-
-      return;
-    }
-
-    stopPolling();
+  shouldSubscribe() {
+    return this.listeners(MESOS_STATE_CHANGE).length > 0 && !this.stream;
   }
 
-  shouldPoll() {
-    return this.listeners(MESOS_STATE_CHANGE).length > 0;
+  subscribe() {
+    const mesosStream = container.get(MesosStreamType);
+    const parsers = new Pipeline(Object.values(mesosStreamParsers));
```

Comment:
Is Pipeline a utility just for this MesosStateStore adaptor?

Reply:
Currently yes, the idea was how can we refactor gigantic `switch` conditionals. Pipeline, you can think of it as the unix pipe operator, just passes the output of one function to  the input of the following. When each function decides whether it transforms the input or not.